Modules/_asynciomodule.c

```diff
@@ -2052,7 +2063,9 @@ unregister_task(asyncio_state *state, TaskObj *task)
     task->next->prev = task->prev;
     if (task->prev == NULL) {
         assert(state->asyncio_tasks.head == task);
+        PyMutex_Lock(&state->mutex);
```

Comment:
Locking needs to cover from before `task->next->prev = task->prev` to after the `else` statement.